README
== About ==
This is a screen scraper for the FFXIV community side, Lodestone. It gets information about
characters. The Lodestone pages are a huge pile of shit, so this will save you a few hours of
hating your life while you write a screen scraper.

The aim is to not break if SquareEnix adds new pieces of data to the page. Any new attributes,
skills, of general pieces of profile data should automatically be handled by this library.

She ain't pretty, but it's decently robust. 

== Installation ==
It's easy:

  $ gem install ffxiv-lodestone

You will need to install the json (pure or extension) gem too. This isn't a dependency because I
don't know of a way to specify an either-or (json-pure or json the native extension) dependency.

== Usage ==
Take, for example, this character: <http://lodestone.finalfantasyxiv.com/rc/character/status?cicuid=1502635>

In its most simplistic instantiation, Character.new with the cicuid from that URL:

  require 'rubygems'
  require 'ffxiv-lodestone'
  char = FFXIVLodestone::Character.new(1502635)

However, you can also search by name and world. If you're doing name/world, you MUST specify both
of these options of you will get an exception.

  char = FFXIVLodestone::Character.new(:name => 'Ayeron Lifebloom', :world => 'Figaro')

:siren: THERE ARE A NUMBER OF CAVEATS INVOLVED IN LOADING-BY-NAME :siren:

  1. This involves two HTTP GETs - one for the search page, and one for the profile page. If you
     are integrating this in to forums or something, you ought to store the character_id number after
     your first search and use that for subsequent requests.

  2. Lodestone's character search is slow to index new characters. It took about a day for my
     character to show up.

  3. Character search has no way of specifying that you want to search for an exact name. If you
     look up a character with a simple name, like 'Ab Lo', you get multiple search results. In the
     event that the search has more than one result, an AmbiguousNameError will be raised. You may
     wish to use the search method and allow your user to pick the correct character, then load it
     by ID, as demonstrated by this psuedo-code:

      chars = FFXIVLodestone::Character.search(:name => 'Lady G', :world => 'Figaro')

      if chars.length ==  1
        char_id = chars.first[:id]
      else
        # display both on a screen for the user to pick, then load the right character by ID
        char_id = . . .
      end

      character = FFXIVLodestone::Character.new(char_id)

    Caveat: This will only display the first page of search results (so maximum 40). If the
    character name you are searching for is _that_ ambiguous, then just fucking delete it and
    reroll with a sane name.

== Ghetto API Doc ==
The static search method will return a list of hashes. You can use this to avoid ambiguious name
errors. Each hash in the array will have id, name, world, and portrait_thumb_url.

  results = Character.search(:name => 'Lady G', :world => 'Figaro') # :world is optional, by the way.

You can results.to_json or results.to_yaml this without any problems.

Stuff you can do with a loaded character:

  char.to_json => String containing a JSON representation of the character / jobs.
  char.to_yaml => YAML representation of the character (EXPERIMENTAL).

  char.name, char.world, char.nameday, char.guardian, char.starting_city, char.physical_level,
  char.current_exp, char.exp_to_next_level, char.gender, char.race, char.clan, char.character_id,
  char.portrait_url, char.portrait_thumb_url

  char.jobs => SkillList with these methods:
    .levelled => Array of all skills that have been leveled (ie rank 0 jobs omitted)
    .pugilist (or any class name) => The skill, regardless of if its been leveled or not
              (ie you can explicitly ask for a skill to see if it's been leveled or not)

  char.resistances => StatList (same deal as stats but with fire instead of strength, 
                      see method list below)
  char.stats => StatList with these methods:
    .strength (or any stat name) => Integer value of the state

  A skill object has these methods: 
    name, skill_name, rank, skillpoint_to_next_level, current_skill_points

== FAQ ==
Q. Hey, why doesn't this have HP / MP / TP? They're on the character profile page!
A. Because those values change based on the currently equipped job. I don't think they're useful
   pieces of data. If you have a use case for them, please let me know and I will include them.

Q. I see you have a hard-coded list of server to ID mappings. That's awful! What am I supposed to
   do when new servers open up?
A. Yes, it is quite horrible! Hard-coding the list is less bad than doing another HTTP GET when
   you use Character.search(), though.

   But I've anticipated your problem. If new worlds open, you can just pass :world as the ID from
   the search page's dropdown until I update the gem:

     FFXIVLodestone::Character.search(:name => 'Ayeron Lifebloom', :world => 7)
     FFXIVLodestone::Character.new(:name => 'Ayeron Lifebloom', :world => 7) 

== Contributing ==
If you want to fool around with the source code, you can see if you've fucked everything up:

  ffxiv-lodestone$ gem install bacon
  ffxiv-lodestone$ cd test/
  ffxiv-lodestone$ bacon spec.rb

== TODO ==
* Figure out what the fuck the stats in parenthesis mean and do something with them.
  * Currently, these are dropped on the floor because I don't know what they mean. The current
    theories include 'Base (with gear)' and 'Stat (with buffs)'.
* Perhaps be able to page through #search() results instead of only having the first 40 available.


HISTORY
== 0.9.7 ==
* 'Shield' has now been aliased to 'sentinel' for char#jobs(). Please be aware that the shield
  key will no longer show up in a json'd character's job list - it will be sentinel instead.

* Karnak added to the server name => search ID mapping.

== 0.9.6 == 
* User agent is now Mozilla Firefox's string. This is because Square Enix needs to be dragged
  kicking and screaming in to 2010.

== 0.9.5 ==
* :world is no longer a required argument when loading by name and searching.
* User agent is now 'ffxiv-lodestone-VER (Ruby/RUBY-VER)'.

== 0.9.4 ==
* Can load non-Gridanian characters now. I guess I'd never tested against a character that
  started in Limsa or Uld - there's an ID that's different between the three so the top box can
  be different colours. 

  I am literally the worst developer ever and I apologize most profusely to everyone because holy
  shit that's kind of a massive bug.

== 0.9.3 ==
* Character.search() will now return up to 40 results; previously, it was only 20.

* Character#skills now has an alias - #jobs. Sementically, char.jobs.carpenter makes more sense
  than char.skills.carpenter - the skill name is carpentry.

  They return exactly the same data, but #skills should be considered deprecated. I may replace it
  with something else in the future.

* When specifying :world in Character.search() and Character.new(), you can pass the server's ID
  number instead of the server's name. You should only use the ID if new servers are opened and
  the gem has not been updated with the name => ID mapping yet.

* strip_nbsp and strip_nbsp! are added to String to handle Nokogiri's oddness more cleanly. I hope
  this doesn't fuck up your application.

== 0.9.2 ==
* Added Character.search() method. It wraps the character search page; an array of hashes will
  be returned with the resulting characters. 
* The full-size portrait URL and thumbnail portrait URLs are now scraped in.

== 0.9.1 ==
* Rails compatability! The #to_json() methods now take an argument (which is promptly dropped on
  the floor) to make ActiveSupport happy.

== 0.9.0 ==
* SkillList#list() changed to SkillList#levelled().
* Characters can be searched and loaded by name. See the README for directions on how to do this.
* Refactored, many things now extend hash.


test/spec.rb
$:.unshift("#{File.expand_path(File.dirname(__FILE__))}/../lib")
require 'bacon'
require 'ffxiv-lodestone'

# Make the character class open our saved HTML...
class FFXIVLodestone::Character
  def self.get_profile_html(id)
    open("./characters/#{id}.html")
  end

  def self.get_search_html(character,world_id)
    n = character.downcase.gsub(' ','-')
    open("./searches/#{n}_#{world_id}.html")
  end
end

describe 'Character.search' do
  it 'should raise an argument error' do
    should.raise(ArgumentError) { FFXIVLodestone::Character.search() }  
    should.raise(ArgumentError) { FFXIVLodestone::Character.search(:irrelevant_key => 'value') }  
    should.raise(ArgumentError) { FFXIVLodestone::Character.search(:world => 'Figaro') }  
    should.raise(ArgumentError) { FFXIVLodestone::Character.search(12345) }  
    should.raise(ArgumentError) { FFXIVLodestone::Character.search(:name => 'Ayeron Lifebloom', :world => 'FAKE SERVER NAME') }  
  end

  it 'should accept server as an integer' do
    should.not.raise(ArgumentError) { FFXIVLodestone::Character.search(:name => 'Ayeron Lifebloom', :world => 7) }
    
    c = FFXIVLodestone::Character.new(:name => 'Ayeron Lifebloom', :world => 7)
    c.character_id.should.equal 1502635
  end

  it 'should be empty' do
    FFXIVLodestone::Character.search(:name => 'ABLOO BLOO UGUU', :world => 'Selbina').should.equal([])
  end
  
  it 'should allow no :world' do
    should.not.raise(ArgumentError) { FFXIVLodestone::Character.search(:name => 'Ayeron Lifebloom')}
    FFXIVLodestone::Character.search(:name => 'Ayeron Lifebloom').should.equal([{:portrait_thumb_url=>"http://static.finalfantasyxiv.com/csnap/1drdb_ss_14e13bdee6f804fa6b989297ba747306.png?geepsvty", :world=>"Figaro", :name=>"Ayeron Lifebloom", :id=>1502635}])
  end

  it 'should list characters' do
    FFXIVLodestone::Character.search(:name => 'Lady', :world => 'Selbina').should.equal(
      [{:world=>"Selbina", :portrait_thumb_url=>"http://static.finalfantasyxiv.com/csnap/v05m_ss_7bd793d507a92d2c415b306a83280d19.png?gediwpzz", :name=>"Lady Simmons", :id=>1015990}, {:world=>"Selbina", :portrait_thumb_url=>"http://static.finalfantasyxiv.com/csnap/14fij_ss_f19cd042628445e22a17a9362cb91f26.png?gee0yejc", :name=>"Shukick Fairlady", :id=>1195603}]
    )
  end
end
describe 'Character(invalid)' do 
  it 'is an invalid id' do
    should.raise(FFXIVLodestone::Character::NotFoundException) { FFXIVLodestone::Character.new('invalid') }
  end

  it 'is a bad argument list' do
    should.raise(ArgumentError) { FFXIVLodestone::Character.new(:id => 1, :name => 'Ayeron Lifebloom', :world => 'Figaro') }
    should.raise(ArgumentError) { FFXIVLodestone::Character.new(:world => 'Figaro') }
  end

  it 'is a bad world' do
    should.raise(ArgumentError) { FFXIVLodestone::Character.new(:name => 'Ayeron Lifebloom', :world => 'FAKESERVER') }
  end

  it 'is an ambiguous name' do
    should.raise(FFXIVLodestone::Character::AmbiguousNameError) { FFXIVLodestone::Character.new(:name => 'Lady', :world => 'Selbina') }
  end

  it 'should actually work' do
    should.not.raise(ArgumentError) { FFXIVLodestone::Character.new(:id => 1015990) }
    should.not.raise(ArgumentError) { FFXIVLodestone::Character.new(:name => 'Ayeron Lifebloom', :world => 'Figaro') }
    should.not.raise(ArgumentError) { FFXIVLodestone::Character.new(:name => 'Ayeron Lifebloom') }
  end
end

describe "Character.new(:name => foo, :world => bar)" do
  it 'should find the character' do
    char = FFXIVLodestone::Character.new(:name => 'Lady Simmons', :world => 'Selbina')
    char.name.should.equal 'Lady Simmons'
    char.character_id.should.equal 1015990

    char = FFXIVLodestone::Character.new(:name => 'Ayeron Lifebloom', :world => 'Figaro')
    char.name.should.equal 'Ayeron Lifebloom'
    char.character_id.should.equal 1502635
    char.physical_level.should.equal 12

    char = FFXIVLodestone::Character.new(:id => 2172370)
    char.name.should.equal 'Karen Kranfel'
    char.character_id.should.equal 2172370

    # This character started in LL. The page structure is slightly different than a Gridanian
    # character. Why? So the top box can be red. :psyduck:
    char = FFXIVLodestone::Character.new(:name => 'Ryoko Antihaijin', :world => 'Figaro')
    char.name.should.equal 'Ryoko Antihaijin'
    char.character_id.should.equal 1124548
    char.starting_city.should.equal 'Limsa Lominsa'
    char.jobs.thaumaturge.rank.should.equal 29 # pro-tier

    # For the sake of completeness, here's one that started in Ul'dah.
    char = FFXIVLodestone::Character.new(:id => 2440978)
    char.name.should.equal 'Warukyure Asura'
    char.character_id.should.equal 2440978
    char.world.should.equal 'Trabia'
    char.starting_city.should.equal "Ul'dah"

    # Worldless!
    char = FFXIVLodestone::Character.new(:name => 'Ayeron Lifebloom')
    char.name.should.equal 'Ayeron Lifebloom'
    char.character_id.should.equal 1502635
    char.world.should.equal 'Figaro'
  end
end

describe 'Character(1015990)' do
  before do
    @char = FFXIVLodestone::Character.new(1015990)
  end

  it 'should have profile data' do
    @char.character_id.should.equal 1015990
    @char.name.should.equal 'Lady Simmons'
    @char.world.should.equal 'Selbina'
    @char.physical_level.should.equal 12
    @char.exp_to_next_level.should.equal 28000
    @char.current_exp.should.equal 18880 
    @char.gender.should.equal 'Female'
    @char.race.should.equal "Miqo'te"
    @char.clan.should.equal 'Seeker of the Sun'
    @char.portrait_url.should.equal 'http://static.finalfantasyxiv.com/csnap/v05m_m_7bd793d507a92d2c415b306a83280d19.png'
    @char.portrait_thumb_url.should.equal 'http://static.finalfantasyxiv.com/csnap/v05m_ss_7bd793d507a92d2c415b306a83280d19.png'
  end

  it 'should have stat lists' do
    stats = {:strength => 16, :vitality => 24, :dexterity => 17, :intelligence => 36, :mind => 37, :piety => 28}
    resists = {:fire => 28, :water => 29, :lightning => 30, :wind => 30, :earth => 25, :ice => 15}
    
    @char.stats.should.equal stats
    @char.stats.strength.should.equal 16

    @char.resistances.should.equal resists 
    @char.resistances.fire.should.equal 28
  end

  it 'should have job data' do
    @char.skills.carpenter.rank.should.equal 5
    @char.skills.carpenter.current_skill_points.should.equal 305 
    @char.skills.carpenter.skillpoint_to_next_level.should.equal 1500

    # make sure the alias is working!
    @char.jobs == @char.skills
  end

  it 'should list all leveled jobs' do
    @char.skills.levelled.map(&:name).sort.should.equal ['Conjurer','Carpenter','Leatherworker','Weaver','Alchemist','Miner','Botanist','Fisher'].sort
  end

  it 'should be convertable to json' do
    json = <<LOLHEREDOC
{"world":"Selbina","clan":"Seeker of the Sun","first_name":"Lady","portrait_url":"http:\/\/static.finalfantasyxiv.com\/csnap\/v05m_m_7bd793d507a92d2c415b306a83280d19.png","character_id":1015990,"current_exp":18880,"last_name":"Simmons","nameday":"4th Sun of the 3rd Umbral Moon","exp_to_next_level":28000,"portrait_thumb_url":"http:\/\/static.finalfantasyxiv.com\/csnap\/v05m_ss_7bd793d507a92d2c415b306a83280d19.png","guardian":"Nymeia, the Spinner","physical_level":12,"jobs":{"alchemist":{"rank":8,"current_skill_points":1753,"skillpoint_to_next_level":3200,"name":"Alchemist","skill_name":"Alchemy"},"marauder":{"rank":0,"current_skill_points":0,"skillpoint_to_next_level":0,"name":"Marauder","skill_name":"Axe"},"blacksmith":{"rank":0,"current_skill_points":0,"skillpoint_to_next_level":0,"name":"Blacksmith","skill_name":"Smithing"},"culinarian":{"rank":0,"current_skill_points":0,"skillpoint_to_next_level":0,"name":"Culinarian","skill_name":"Cooking"},"archer":{"rank":0,"current_skill_points":0,"skillpoint_to_next_level":0,"name":"Archer","skill_name":"Archery"},"armorer":{"rank":0,"current_skill_points":0,"skillpoint_to_next_level":0,"name":"Armorer","skill_name":"Armorcraft"},"miner":{"rank":1,"current_skill_points":431,"skillpoint_to_next_level":570,"name":"Miner","skill_name":"Mining"},"lancer":{"rank":0,"current_skill_points":0,"skillpoint_to_next_level":0,"name":"Lancer","skill_name":"Polearm"},"goldsmith":{"rank":0,"current_skill_points":0,"skillpoint_to_next_level":0,"name":"Goldsmith","skill_name":"Goldsmithing"},"botanist":{"rank":9,"current_skill_points":1800,"skillpoint_to_next_level":4300,"name":"Botanist","skill_name":"Botany"},"thaumaturge":{"rank":0,"current_skill_points":0,"skillpoint_to_next_level":0,"name":"Thaumaturge","skill_name":"Thaumaturgy"},"leatherworker":{"rank":1,"current_skill_points":78,"skillpoint_to_next_level":570,"name":"Leatherworker","skill_name":"Leatherworking"},"pugilist":{"rank":0,"current_skill_points":0,"skillpoint_to_next_level":0,"name":"Pugilist","skill_name":"Hand-to-Hand"},"fisher":{"rank":3,"current_skill_points":748,"skillpoint_to_next_level":880,"name":"Fisher","skill_name":"Fishing"},"conjurer":{"rank":1,"current_skill_points":272,"skillpoint_to_next_level":570,"name":"Conjurer","skill_name":"Conjury"},"weaver":{"rank":6,"current_skill_points":1145,"skillpoint_to_next_level":1800,"name":"Weaver","skill_name":"Clothcraft"},"gladiator":{"rank":0,"current_skill_points":0,"skillpoint_to_next_level":0,"name":"Gladiator","skill_name":"Sword"},"carpenter":{"rank":5,"current_skill_points":305,"skillpoint_to_next_level":1500,"name":"Carpenter","skill_name":"Woodworking"},"sentinel":{"rank":0,"current_skill_points":0,"skillpoint_to_next_level":0,"name":"Sentinel","skill_name":"Shield"}},"race":"Miqo'te","resistances":{"earth":25,"ice":15,"fire":28,"water":29,"lightning":30,"wind":30},"starting_city":"Gridania","attributes":{"mind":37,"piety":28,"strength":16,"vitality":24,"dexterity":17,"intelligence":36},"gender":"Female"}
LOLHEREDOC

    JSON.parse(@char.to_json).should.equal JSON.parse(json)
  end
end


ffxiv_lodestone.rb
require 'rubygems'
require 'nokogiri'
require 'open-uri'
require 'json'

module FFXIVLodestone
  class Character
    class StatList
      include Enumerable 
      attr_reader :stats

      def initialize(table)
        @stats = {}
        
        table.search('tr').each do |tr|
          @stats[tr.children[0].content.strip.downcase.to_sym] = tr.children[2].content.gsub("\302\240",' ').split(' ')[0].to_i
        end
      end

      def each
        @stats.each {|stat| yield stat}
      end

      def to_h
        @stats
      end

      def method_missing(method)
        return @stats[method] if @stats.key? method
        super
      end
    end # StatList

    class SkillList
      class Skill
        attr_reader :name, :skill_name, :rank, :current_skill_points, :skillpoint_to_next_level

        def initialize(job,skill_name,rank,cur_sp,skillup_sp)
          @name = job 
          @skill_name = skill_name
          @rank = rank
          @current_skill_points = cur_sp
          @skillpoint_to_next_level = skillup_sp
        end # initalize

        def to_h
          {:name => @name, :skill_name => @skill_name, :rank => @rank, :current_skill_points => @current_skill_points, :skillpoint_to_next_level => @skillpoint_to_next_level}
        end
      end # Skill
      
      # Alias the stupid names in Lodestone to class names. 
      SKILL_TO_CLASS = {
        'Hand-to-Hand' => :pugilist,
        'Sword' => :gladiator,
        'Axe' => :marauder,
        'Archery' => :archer,
        'Polearm' => :lancer,
        'Thaumaturgy' => :thaumaturge,
        'Conjury' => :conjurer,
        'Woodworking' => :carpenter,
        'Smithing' => :blacksmith,
        'Armorcraft' => :armorer,
        'Goldsmithing' => :goldsmith,
        'Leatherworking' => :leatherworker,
        'Clothcraft' => :weaver,
        'Alchemy' => :alchemist,
        'Cooking' => :culinarian,
        'Mining' => :miner,
        'Botany' => :botanist,
        'Fishing' => :fisher,
      }

      def initialize(skill_table)
        @skills = {}

        skill_table.children.each do |skill|
          name = skill.children[0].children[1].content
          if SKILL_TO_CLASS.key? name
            key = SKILL_TO_CLASS[name]
            job = key.to_s.capitalize
          else
            key = name.gsub('-', '_').downcase.to_sym
            job = name
          end

          # '-' = not leveled (never equipped the class' weapon)
          rank = skill.children[2].search('table tr td[last()]').children.first.content
          rank = (rank.include?('-') ? 0 : rank.to_i)

          # # '-' = not leveled, otherwise it will be in the format '391 / 1500'
          sp = skill.children[4].search('table tr td[last()]').children.first.content

          if sp.include? '-'
            current_sp = 0
            levelup_sp = 0
          else
            sp.gsub!("\302\240",'') # this is a &nbsp but it looks ugly in #inspect(), so remove it.
            current_sp = sp.split('/')[0].strip.to_i
            levelup_sp = sp.split('/')[1].strip.to_i
          end

          @skills[key] = Skill.new(job,name,rank,current_sp,levelup_sp)
        end
      end # initialize

      # Lists all leveled jobs.
      def list
        list = []
        @skills.each do |name,skill| 
          list << skill if skill.rank > 0
        end

        return list
      end

      def to_h
        list = {}
        @skills.each {|job,data| list[job] = data.to_h}

        return list
      end
      
      def method_missing(method)
        return @skills[method] if @skills.key? method
        super
      end

    end # FFXIVLodestone::Character::SkillList

    attr_reader :skills, :stats, :resistances, :profile
    def initialize(character_id)
      # TODO exception if ID isn't an integer
      # TODO exception if we don't have a valid char ID
      @character_id = character_id

      doc = Nokogiri::HTML(open("http://lodestone.finalfantasyxiv.com/rc/character/status?cicuid=#{@character_id}", {'Accept-Language' => 'en-us,en;q=0.5', 'Accept-Charset' => 'utf-8;q=0.5'}))

      # The skills table doesn't have a unqiue ID or class to find it by, so take the first skill lable and go up two elements (table -> tr -> th.mianskill-lable)
      @skills = SkillList.new(doc.search('th.mainskill-label').first.parent.parent)
      @stats = StatList.new(doc.search("div.contents-subheader[contains('Attributes')]").first.next_sibling.next_sibling)
      @resistances = StatList.new(doc.search("div.contents-subheader[contains('Elements')]").first.next_sibling.next_sibling)
      
      # The character info box at the top ... actually has a useful ID!
      @profile = {}
      profile = doc.search('#profile-plate2')
      profile.search('tr th').each do |th|
        key = th.content.strip.downcase.gsub(':','').gsub(' ','_').to_sym
        value = th.next_sibling.next_sibling.content.strip.gsub("\302\240",'')

        # HP/MP/TP are max values. They depend on the currently equipped job and are not very
        # meaningful pieces of data. XP will be handled seperately. 
        unless [:hp, :mp, :tp, :experience_points].include? key
          @profile[key] = value 
        end

        if key == :experience_points
          @profile[:current_exp] = value.split('/')[0].to_i
          @profile[:exp_to_next_level] = value.split('/')[1].to_i
        end
      end
      
      # Fix this datatype.
      @profile[:physical_level] = @profile[:physical_level].to_i
      
      # Parse the character name/world line...
      name_line = profile.search('#charname').first.content.gsub(')','').strip.split(' (')
      @profile[:world] = name_line[1]
      @profile[:first_name] = name_line[0].split(' ')[0]
      @profile[:last_name] = name_line[0].split(' ')[1]

      # Parse the "Seeker of the Sun Female / Miqo'te" line... fun~
      race_line = profile.search('tr td').first.content.strip.gsub("\302\240",' ').split(' / ')
      @profile[:race] = race_line.pop

      # horrible array splitting and popping trix. hidoi hidoi!
      race_line = race_line.first.split ' '
      @profile[:gender] = race_line.pop
      @profile[:clan] = race_line.join ' '
    end
    
    def name
      "#{@profile[:first_name]} #{@profile[:last_name]}"
    end

    def to_json
      data = {}
      data.merge!(@profile)
      data[:jobs] = @skills.to_h
      data[:attributes] = @stats.to_h
      data[:resistances] = @resistances.to_h

      data.to_json
    end

    def method_missing(method)
      return @profile[method] if @profile.key? method
      super
    end
  end # character
end # end FFXIVLodestone


lib/ffxiv-lodestone.rb
require 'rubygems'
require 'nokogiri'
require 'open-uri'
require 'json'

# Nokogiri changes &nbsp; entities to \302\240. This may be OK in most instances,
# but at the very least, it makes #inspect() look like shit, which is a headache
# for people trying to troubleshoot.
class String
  def strip_nbsp
    self.strip.gsub("\302\240",' ')
  end

  def strip_nbsp!
    before = self.reverse.reverse
    self.strip!
    self.gsub!("\302\240",' ')
    before == self ? nil : self
  end
end

module FFXIVLodestone
  # Gem version.
  VERSION = '0.9.7'

  # Accept-language must be sent; their default is Japanese text.
  HTTP_OPTIONS = {'Accept-Language' => 'en-us,en;q=0.5', 'Accept-Charset' => 'utf-8;q=0.5', 
    'User-Agent' => "Mozilla/5.0 (Windows; U; Windows NT 6.1; en-US; rv:1.9.2.08) Gecko/20100914 Firefox/3.6.10"}
  
  # Search page server IDs.
  SERVER_SEARCH_INDEXES = {:cornelia => 2, :kashuan => 3, :gysahl => 4, :mysidia => 5, 
  :istory => 6, :figaro => 7, :wutai => 8, :trabia => 9, :lindblum => 10, :besaid => 11,
  :selbina => 12, :rabanastre => 13, :bodhum => 14, :melmond => 15, :palamecia => 16,
  :saronia => 17, :fabul => 18, :karnak => 19}
  
  # Alias the stupid names in Lodestone to class names. 
  SKILL_TO_CLASS = {
    'Hand-to-Hand' => :pugilist,
    'Sword' => :gladiator,
    'Axe' => :marauder,
    'Archery' => :archer,
    'Polearm' => :lancer,
    'Thaumaturgy' => :thaumaturge,
    'Conjury' => :conjurer,
    'Woodworking' => :carpenter,
    'Smithing' => :blacksmith,
    'Armorcraft' => :armorer,
    'Goldsmithing' => :goldsmith,
    'Leatherworking' => :leatherworker,
    'Clothcraft' => :weaver,
    'Alchemy' => :alchemist,
    'Cooking' => :culinarian,
    'Mining' => :miner,
    'Botany' => :botanist,
    'Fishing' => :fisher,
    'Shield' => :sentinel,
  }

  module Serializable
    def to_yaml_properties
      serialize_properties.map {|sym| "@%s" % sym }
    end

    def to_hash
      serialize_properties.reduce(Hash.new) {|h,sym|
        h[sym] = self.send(sym)
        h
      }
    end
    alias :to_h :to_hash # legacy API

    def to_json(args={})
      self.to_hash.to_json
    end
  end

  # Represents an FFXIV character. Example:
  #
  #   FFXIVLodestone::Character.new(1015990)
  class Character
    class NotFoundException < RuntimeError 
    end

    class AmbiguousNameError < RuntimeError
    end

    class StatList < Hash
      def initialize(table)
        table.search('tr').each do |tr|
          self[tr.children[0].content.strip.downcase.to_sym] = tr.children[2].content.strip_nbsp.split(' ')[0].to_i
        end
      end

      def method_missing(method)
        return self[method] if self.key? method
        super
      end
    end # StatList

    class SkillList < Hash 
      class Skill
        include Serializable

        attr_reader :name, :skill_name, :rank, :current_skill_points, :skillpoint_to_next_level
        def initialize(job,skill_name,rank,cur_sp,skillup_sp)
          @name = job 
          @skill_name = skill_name
          @rank = rank
          @current_skill_points = cur_sp
          @skillpoint_to_next_level = skillup_sp
        end # initalize

        def serialize_properties
          [:name, :skill_name, :rank, :current_skill_points, :skillpoint_to_next_level]
        end
      end # Skill
      
      def initialize(skill_table)
        @skills = {}

        skill_table.children.each do |skill|
          name = skill.children[0].children[1].content
          if FFXIVLodestone::SKILL_TO_CLASS.key? name
            key = FFXIVLodestone::SKILL_TO_CLASS[name]
            job = key.to_s.capitalize
          else
            key = name.gsub('-', '_').downcase.to_sym
            job = name
          end

          # '-' = not leveled (never equipped the class' weapon)
          rank = skill.children[2].search('table tr td[last()]').children.first.content
          rank = (rank.include?('-') ? 0 : rank.to_i)

          # # '-' = not leveled, otherwise it will be in the format '391 / 1500'
          sp = skill.children[4].search('table tr td[last()]').children.first.content

          if sp.include? '-'
            current_sp = 0
            levelup_sp = 0
          else
            sp.strip_nbsp!
            current_sp = sp.split('/')[0].strip.to_i
            levelup_sp = sp.split('/')[1].strip.to_i
          end

          self[key] = Skill.new(job,name,rank,current_sp,levelup_sp)
        end
      end # initialize

      # Lists all leveled (rank > 0) jobs.
      def levelled 
        self.keys.reduce(Array.new) do |a,key| 
          a << self.send(key) if self.send(key).rank > 0
          a
        end
      end

      def method_missing(method)
        return self[method] if self.key? method
        super
      end
    end # FFXIVLodestone::Character::SkillList

    attr_reader :skills, :stats, :resistances, :profile
    alias :jobs :skills
    def initialize(args={})
      args = {:id => args} unless args.class == Hash
      raise ArgumentError, 'No search paremeters were given.' if args.empty?

      if args.key? :id
        character_id = args[:id]
        raise ArgumentError, 'No other arguments may be specified in conjunction with :id.' if args.size > 1
      else
        characters = Character.search(args)
        raise NotFoundException, 'Character search yielded no results.' if characters.empty?
        raise AmbiguousNameError, 'Multiple characters matched that name.' if characters.size > 1
        character_id = characters.first[:id]
      end
      
      @character_id = character_id
      doc = Nokogiri::HTML(Character.get_profile_html(@character_id))

      # Did we get an error page? Invalid ID, etc.
      if !((doc.search('head title').first.content.match /error/i) == nil)
        raise NotFoundException, 'Bad character ID or Lodestone is broken.' 
      end

      # The skills table doesn't have a unqiue ID or class to find it by, so take the first skill lable and go up two elements (table -> tr -> th.mianskill-lable)
      @skills = SkillList.new(doc.search('th.mainskill-label').first.parent.parent)
      @stats = StatList.new(doc.search("div.contents-subheader[contains('Attributes')]").first.next_sibling.next_sibling)
      @resistances = StatList.new(doc.search("div.contents-subheader[contains('Elements')]").first.next_sibling.next_sibling)
      
      # The character info box at the top ... actually has a useful ID!
      @profile = {}
      profile = doc.search("//div[starts-with(@id,'profile-plate')]") 
      profile.search('tr th').each do |th|
        key = th.content.strip.downcase.gsub(':','').gsub(' ','_').to_sym
        value = th.next_sibling.next_sibling.content.strip_nbsp

        # HP/MP/TP are max values. They depend on the currently equipped job and are not very
        # meaningful pieces of data. XP will be handled seperately. 
        unless [:hp, :mp, :tp, :experience_points].include? key
          @profile[key] = value 
        end

        if key == :experience_points
          @profile[:current_exp] = value.split('/')[0].to_i
          @profile[:exp_to_next_level] = value.split('/')[1].to_i
        end
      end
      
      # Fix this datatype.
      @profile[:physical_level] = @profile[:physical_level].to_i
      
      # Parse the character name/world line...
      name_line = profile.search('#charname').first.content.gsub(')','').strip.split(' (')
      @profile[:world] = name_line[1]
      @profile[:first_name] = name_line[0].split(' ')[0]
      @profile[:last_name] = name_line[0].split(' ')[1]

      # Parse the "Seeker of the Sun Female / Miqo'te" line... fun~
      race_line = profile.search('tr td').first.content.strip_nbsp.split(' / ')
      @profile[:race] = race_line.pop

      # horrible array splitting and popping trix. hidoi hidoi!
      race_line = race_line.first.split ' '
      @profile[:gender] = race_line.pop
      @profile[:clan] = race_line.join ' '

      @profile.merge! generate_portrait_urls(doc.search('div.image-mount-image img').first.attr('src').strip)

      @profile[:character_id] = @character_id.to_i
    end

    # FFXIVLodestone::Character.search(:name => 'Character Name', :world => 'Server') => Array
    def self.search(args={})
      raise ArgumentError, 'Search parameters must be hash.' unless args.class == Hash
      raise ArgumentError, ':name must be specified to use search.' unless args.key? :name

      world_id = nil
      if args.key? :world
        # :world can be passed as a string ('Figaro') or as the integer used by the search page (7).
        # This is so the library is not completely useless when new worlds are added - developers can
        # fall back to the integers until the gem is updated.
        if args[:world].class == String
          raise ArgumentError, 'Unknown world server.' unless FFXIVLodestone::SERVER_SEARCH_INDEXES.key? args[:world].downcase.to_sym 

          world_id = FFXIVLodestone::SERVER_SEARCH_INDEXES[args[:world].downcase.to_sym] 
        else
          world_id = args[:world].to_i # force it to an int to prevent any funny business.
        end
      end

      doc = Nokogiri::HTML(get_search_html(args[:name],world_id))
      results = doc.search("table.contents-table1 tr th[contains('Character Name')]")
      return [] if results.empty? # No results = no results table header. 
      
      results = results.last.parent.parent
      results.children.first.remove # discard the table headers
      
      results.children.map do |tr|
        name_element = tr.search('td:first table tr td:last a').first
        {
          :id => name_element.attr('href').gsub('/rc/character/top?cicuid=','').strip.to_i,
          :name => name_element.content.strip,
          :portrait_thumb_url => tr.search('td:first table tr td:first img').first.attr('src').strip,
          :world => tr.search('td:last').last.content.strip
        }
      end
    end # search
    
    # Returns first name / last name seperated by a space.
    def name
      "#{@profile[:first_name]} #{@profile[:last_name]}"
    end

    def to_json(args={})
      data = {}
      data.merge!(@profile)
      data[:jobs] = @skills
      data[:attributes] = @stats
      data[:resistances] = @resistances

      data.to_json
    end

    def method_missing(method)
      return @profile[method] if @profile.key? method
      super
    end

    protected 
    # This method can be redefined in a test class.
    def self.get_profile_html(id)
      open("http://lodestone.finalfantasyxiv.com/rc/character/status?cicuid=#{id}", FFXIVLodestone::HTTP_OPTIONS)
    end

    # Another method to redefine in the test file...
    def self.get_search_html(name,world_id)
      open(URI.encode("http://lodestone.finalfantasyxiv.com/rc/search/search?tgt=77&q=#{name}&cw=#{world_id}&num=40"), FFXIVLodestone::HTTP_OPTIONS)
    end

    def generate_portrait_urls(url)
      url_list = {}

      a = URI.parse(url)
      a.query = nil
      url_list[:portrait_url] = a.to_s

      a.path = a.path.gsub(/\/([A-Z0-9]+)_m_/i,'/\1_ss_')
      url_list[:portrait_thumb_url] = a.to_s

      return url_list
    end
  end # character
end # end FFXIVLodestone
